modules/caddyhttp/fileserver/matcher.go

```diff
@@ -183,55 +179,79 @@ func (m MatchFile) selectFile(r *http.Request) (rel, abs string, matched bool) {
 		m.TryFiles = []string{r.URL.Path}
 	}
 
+	// common preparation of the file into parts
+	prepareFilePath := func(file string) (string, string) {
+		suffix := m.firstSplit(path.Clean(repl.ReplaceAll(file, "")))
+		if strings.HasSuffix(file, "/") {
+			suffix += "/"
+		}
+		fullpath := sanitizedPathJoin(root, suffix)
+		return suffix, fullpath
+	}
+
+	// sets up the placeholders for the matched file
+	setPlaceholders := func(info os.FileInfo, rel string, abs string) {
+		repl.Set("http.matchers.file.relative", rel)
+		repl.Set("http.matchers.file.absolute", abs)
+
+		fileType := "file"
+		if info.IsDir() {
+			fileType = "directory"
+		}
+		repl.Set("http.matchers.file.type", fileType)
+	}
+
 	switch m.TryPolicy {
 	case "", tryPolicyFirstExist:
 		for _, f := range m.TryFiles {
-			suffix := m.firstSplit(path.Clean(repl.ReplaceAll(f, "")))
-			fullpath := sanitizedPathJoin(root, suffix)
-			if strictFileExists(fullpath) {
-				return suffix, fullpath, true
+			suffix, fullpath := prepareFilePath(f)
+			if exists, info := strictFileExists(fullpath); exists {
```

Comment:
Usually the second return value would be the bool, indicating whether the first is usable. However, `info` can be nil, so we can use that as the signal for whether it matched or not.

```suggestion
			if info := strictFileExists(fullpath); info != nil {
```

I think this is simpler and equivalent. Can we rework it to be like this?

Reply:
Mmm yeah that works :+1:

Reply:
I just flipped the return values instead, the condition logic would be kinda funky :nose: if trying to make `info != nil` the check